Add how to change image shown on an Image element

rollersteaam opened the pull request:
**What kind of change does this PR introduce?**
Adds information about how to change the image shown on an Image element.

**What is the current behavior?**
There is no information about how binding to an Image's `Source` property doesn't work, and how to fix that problem.

https://github.com/AvaloniaUI/Avalonia/issues/3774
https://github.com/AvaloniaUI/Avalonia/issues/853
https://github.com/AvaloniaUI/Avalonia/issues/3860

**What is the new behavior?**
That information now exists, and an example for a declarative approach in the form of a "Swappable Image Button" has been made.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/AvaloniaUI/avaloniaui.net?pullRequest=169) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/avaloniaui.net?pullRequest=169) before we can accept your contribution.<br/><br/>:x: rollersteaam [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/avaloniaui.net?pullRequest=169)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/avaloniaui.net?pullRequest=169) it.</sub>

rollersteaam commented:
@jmacato I wanted to give credit to you but it feels a bit awkward doing that in documentation. This example is adapted based off of what you gave in https://github.com/AvaloniaUI/Avalonia/issues/3860 though.

grokys commented:
This seems like a good thing to have an example of in the docs, but also I seem to remember you were speaking with jumar in chat and came up with a better way to do this using `ToggleButton` and styles, am I correct?

If so, maybe this would be better living in the `ToggleButton`  docs?

rollersteaam commented:
> 
> 
> This seems like a good thing to have an example of in the docs, but also I seem to remember you were speaking with jumar in chat and came up with a better way to do this using `ToggleButton` and styles, am I correct?
> 
> If so, maybe this would be better living in the `ToggleButton` docs?

Good thoughts, in fact I think adding examples in Button and ToggleButton would be equally beneficial, and then adding a "Source Binding" example in Image would be the icing on the cake. Will do that now.

rollersteaam commented:
Done. I've added new doc pages for `ToggleButton` and `DrawingPresenter` because they weren't there, as well as noted that they are subclasses in the relevant parent class pages (Button and Image). I've also added the binding converter approach and `ToggleButton` approaches to the Image page, and added a "Speaker Mute Button" example in the `ToggleButton` page, and a "Play Button" example in the `Button` page.

rollersteaam commented:
Thank you to @jmacato and @ahopper for your respective code examples, wouldn't be possible without your help and knowledge. And thanks to @define-private-public for creating https://github.com/AvaloniaUI/Avalonia/issues/3860 where this could be discussed.

define-private-public commented:
In the PR's text, can you also add an example of what you should do in the `<XXX.Resources>` block for an `<Image>` element?  I see `Source={DynamicResource Play}`,  but I don't see the key `Play` defined.

grokys commented on src/AvaloniaUI.Net/wwwroot/docs/controls/drawingpresenter.md:
```diff
@@ -0,0 +1,34 @@
+---
+Title: DrawingPresenter
+---
+The `DrawingPresenter` control is a subclass of the `Image` control, which can be used to display svgs.
```
Not true: `DrawingPresenter` does not derive from `Image`. From 0.10.0 `DrawingPresenter` will no longer be necessary (will be deprecated) and one should just use `Image`, so I think we should probably remove the new docs from the file as they will become obsolete in the next few weeks.

rollersteaam replied:
DrawingPresenter now removed, not worth adding if it's going to be marked as deprecated anyway. https://github.com/AvaloniaUI/avaloniaui.net/pull/169/commits/ad9c7ffff576b38844065bfb47dacbe3d75d7402

grokys commented on src/AvaloniaUI.Net/wwwroot/docs/controls/image.md:
```diff
@@ -3,5 +3,121 @@ Title: Image
 ---
 The `Image` control is a control for displaying raster images.
 
+Binding on the `Image` control's `Source` property will not work unless you use a [binding converter](https://avaloniaui.net/docs/binding/converting-binding-values) that will convert a `string` into a `IBitmap`.
```
Not true: binding an image control's `Source` property to an `IBitmap` will work fine. The converter is only needed when binding `Source` to a `string`.

rollersteaam replied:
Information corrected https://github.com/AvaloniaUI/avaloniaui.net/pull/169/commits/454a8b154454419b366c2b5e841c2c1299257046, I've also changed the tone to be more helpful rather than complaining. It's also in its own heading now to catch the eyes of anyone skimming the page to find out why binding a string isn't working for them.

grokys commented on src/AvaloniaUI.Net/wwwroot/docs/controls/image.md:
```diff
@@ -3,5 +3,121 @@ Title: Image
+Binding on the `Image` control's `Source` property will not work unless you use a [binding converter](https://avaloniaui.net/docs/binding/converting-binding-values) that will convert a `string` into a `IBitmap`.
+
+## Subclasses
+1. [DrawingPresenter](https://avaloniaui.net/docs/controls/drawingpresenter) - Used to display svgs.
```
Not true: `DrawingPresenter` does not derive from `Image`.

rollersteaam replied:
Information removed in https://github.com/AvaloniaUI/avaloniaui.net/pull/169/commits/04480eec0294438ef315c14c14b4e8db54774565.

grokys commented on src/AvaloniaUI.Net/wwwroot/docs/controls/drawingpresenter.md:
````diff
@@ -0,0 +1,34 @@
+
+### Speaker Mute Button
+```
+<Style Selector="ToggleButton DrawingPresenter.tbchecked">
````
Some explanation of what this example is showing might be useful here.

rollersteaam replied:
Resolved due to the removal of `DrawingPresenter`.

grokys commented on src/AvaloniaUI.Net/wwwroot/docs/controls/image.md:
````diff
@@ -3,5 +3,121 @@ Title: Image
+       Source="{Binding PlaySource, Converter={StaticResource variableImage}}">
+```
+
+#### Declarative Approaches
````
These examples are already provided in the `Button` and `ToggleButton` docs, and they're not particularly relevant to `Image` itself: might be best to just link to them instead of copy/pasting them here?

rollersteaam replied:
Agreed, I was a bit anxious when I realised I was copy and pasting example code on multiple pages. The trade-off being that people don't need to navigate to a different page to know the information. I think links are a better strategy however, and so the examples have been replaced with links to the respective pages and headings https://github.com/AvaloniaUI/avaloniaui.net/pull/169/commits/498686e8fedc9b8538e5b9609295c504a1c8009a.

grokys commented on src/AvaloniaUI.Net/wwwroot/docs/controls/togglebutton.md:
````diff
@@ -0,0 +1,34 @@
+## Examples
+
+### Speaker Mute Button
+```
````
Again, a description of what this is doing would be useful.

rollersteaam replied:
I've added a brief description under the heading that clarifies the point of the example https://github.com/AvaloniaUI/avaloniaui.net/pull/169/commits/0432b29e65a45c0edadcebea19869fb582f99f64. I've also added explanation for the style code, how it works and why https://github.com/AvaloniaUI/avaloniaui.net/pull/169/commits/8f066d19e9f363f745c0d662e357e5bda5633faa.

rollersteaam commented:
> 
> 
> In the PR's text, can you also add an example of what you should do in the `<XXX.Resources>` block for an `<Image>` element? I see `Source={DynamicResource Play}`, but I don't see the key `Play` defined.

I've added some information in the example about how one can setup dynamic resources for `Source` on an `Image` control https://github.com/AvaloniaUI/avaloniaui.net/pull/169/commits/9eaddbf98b074abaa4bc89bdb1f4d505a18aaf13. I hope this code is correct?

rollersteaam commented:
Thank you for the PR review @grokys, it was very helpful. This improves the quality of what I've done significantly.

After this I'll probably be moving onto improving the docs for authoring new controls, since components are a fundamental element of making software that stops projects from spiralling out of control, and it took me 2 days or so to learn and get working, much like changing the images on the Image control haha

grokys commented:
Merged! Sorry for the delay, vacation season got in the way. Your contributions are much appreciated!
